txscript/sigcache.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2015 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package txscript
+
+import (
+	"bytes"
+	"crypto/rand"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// sigInfo represents an entry in the SigCache. Entries in the sigcache are a
+// 3-tuple: (sigCache, sig, pubKey).
+type sigInfo struct {
+	// TODO(roasbeef): Make this a string so it's hash func blocksize
+	// generic?
+	sigHash wire.ShaHash
+	sig     btcec.Signature
```

Comment:
This will not work.  Both `btcec.Signature` and `btcec.PublicKey` contain `big.Int` pointers.  This means two signatures and/or pubkeys that are otherwise identical will not match because the pointers are pointing to different memory.  To prove it, I did the following:

```diff
diff --git a/txscript/sigcache_test.go b/txscript/sigcache_test.go
index 6a1fda6..34375b0 100644
--- a/txscript/sigcache_test.go
+++ b/txscript/sigcache_test.go
@@ -49,7 +49,8 @@ func TestSigCacheAddExists(t *testing.T) {
        sigCache.Add(msg1, sig1, key1)

        // The previously added triplet should now be found within the sigcache.
-       if !sigCache.Exists(msg1, sig1, key1) {
+       sig1Copy, _ := btcec.ParseSignature(sig1.Serialize(), btcec.S256())
+       if !sigCache.Exists(msg1, sig1Copy, key1) {
                t.Errorf("previously added item not found in signature cache")
        }
 }
```
> $ go test
> --- FAIL: TestSigCacheAddExists (0.00s)
> 	sigcache_test.go:57: previously added item not found in signature cache